src/TensorTwoD.hpp

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef TENSORTWOD_HPP_INCLUDE
+#define TENSORTWOD_HPP_INCLUDE
+
+#include <cstddef>
+
+#include "geopm/json11.hpp"
+
+#include "TensorOneD.hpp"
+
+namespace geopm
+{
+
+    class TensorTwoD
+    {
+        public:
+            TensorTwoD();
+            TensorTwoD(std::size_t rows, std::size_t cols);
+            TensorTwoD(const TensorTwoD&);
+            TensorTwoD(json11::Json input);
+            void set_dim(std::size_t rows, std::size_t cols);
+            std::size_t get_rows() const;
+            std::size_t get_cols() const;
+            TensorOneD operator*(const TensorOneD&);
+            TensorOneD &operator[](int idx);
+            TensorOneD operator[](int idx) const;
+            TensorTwoD& operator=(const TensorTwoD&);
+
+        private:
+            std::vector<TensorOneD> m_mat;
```

Comment:
Are you *sure* you want to represent your 2D matrix class as a vector of 1D matrix class? It's basically a vector of vectors, which isn't very cache efficient.

Perhaps you would be willing to try my cache efficient matrix recipe? This ensures that all the values of the matrix are spatially close together, instead of being scattered all over the memory, which makes for a very cache efficient.
```cpp
size_t (*m_mat)[cols] = static_cast<int(*)[cols]>( malloc(sizeof(int[rows][cols])) );

// do whatever you want with it

delete[] m_mat;
```
I can help you rewrite the copy and resize operations on such a data structure, if you would decide to go along with my suggestion.

Reply:
From the discussion yesterday, leaving this as a potential future optimization. The matrices are relatively small, so the operations should not take too long. However, assessing the overhead should be part of future work.